experimental/unitsconversion/src/converter.rs

```diff
@@ -0,0 +1,362 @@
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{
+    measureunit::{MeasureUnit, MeasureUnitParser},
+    provider::{Base, MeasureUnitItem, SiPrefix, Sign, SignULE, UnitsInfoV1Marker},
+    ConversionError,
+};
+
+use icu_provider::DataError;
+use icu_provider::DataPayload;
+use icu_provider::DataProvider;
+use icu_provider::DataRequest;
+use litemap::LiteMap;
+use num::{rational::Ratio, BigInt, Zero};
+use zerotrie::ZeroTrieSimpleAscii;
+use zerovec::{ule::AsULE, ZeroSlice};
+
+/// Represents the possible cases for the convertibility between two units.
+#[derive(Debug, PartialEq)]
+enum Convertibility {
+    /// The units are convertible.
+    /// For example, `meter` and `foot` are convertible.
+    Convertible,
+
+    /// The units are reciprocal.
+    /// For example, `gallon-per-mile` and `100-kilometer-per-liter` are reciprocal.
+    Reciprocal,
+
+    /// The units are not convertible.
+    /// For example, `meter` and `second` are not convertible.
+    NotConvertible,
+}
+
+/// A converter for converting between two single units.
+/// For example: 1- `meter` to `foot`.
+///              2- `square-meter` to `square-foot`.
+///              3- `100-kilometer-per-liter` to `gallon-per-mile`.
+/// NOTE:
+///    The converter is not able to convert between two units that are not single. such as "foot-and-inch" to "meter".
+#[derive(Debug)]
+pub struct Converter {
+    /// The conversion rate between the input and output units.
+    conversion_rate: Ratio<BigInt>,
+
+    /// The offset between the input and output measurements.
+    /// For example, converting between `celsius` and `fahrenheit` requires an offset of 32.
+    offset: Ratio<BigInt>,
+
+    /// Determines if the units are reciprocal or not.
+    /// For example, `meter-per-second` and `second-per-meter` are reciprocal.
+    reciprocal: bool,
+}
+
+/// A factory for creating a converter.
+/// Also, it can check the convertibility between two units.
+pub struct ConverterFactory {
+    // TODO(#4522): Make the converter factory owns the data.
+    /// Contains the necessary data for the conversion factory.
+    /// DataPayload<ErasedPluralRulesV1Marker>
+    data: DataPayload<UnitsInfoV1Marker>,
+    // payload_store: ZeroTrieSimpleAscii<ZeroVec<'static, u8>>,
+}
+
+impl ConverterFactory {
+    #[cfg(feature = "datagen")]
+    pub fn new(data: &(impl DataProvider<UnitsInfoV1Marker> + ?Sized)) -> Result<Self, DataError> {
+        let data = data.load(DataRequest::default())?.take_payload()?.cast();
+
+        // let store: &UnitsInfoV1<'static> = data.get();
+        // let store = store.units_conversion_trie.clone().take_store();
+        Ok(Self { data })
+    }
+
+    pub fn parser(&self) -> MeasureUnitParser<'_> {
+        let trie = self.data.get().units_conversion_trie.clone(); // cheap since store is a borrowed ZeroVec
```

Comment:
Suggestion: it would be better here to borrow from the short-lived reference via `impl ZeroFrom`

https://docs.rs/zerotrie/0.1.2/zerotrie/struct.ZeroTrieSimpleAscii.html#impl-ZeroFrom%3C'zf,+ZeroTrieSimpleAscii%3CStore1%3E%3E-for-ZeroTrieSimpleAscii%3CStore2%3E

Something like this

```rust
let trie = ZeroTrieSimpleAscii::<ZeroVec<u8>>::zero_from(self.data.get().units_conversion_trie)
```

Note: the clone is no-alloc if the data is borrowed but it is not free if the data is owned, but ZeroFrom is always no-alloc.

Reply:
There is no clone right now, I have deleted it :)